File: scripts/spec_fuzz/classify.ts
// Groups the failures of a spec check report by what differs first, so that
// thousands of them can be read as the handful of causes they have. Run it with
// `deno run -A scripts/spec_fuzz/classify.ts <report.txt> parse|format [shown]`.
//
// `parse` reads the examples whose parse differs from the expected html,
// `format` the ones that parsed as expected but whose formatted output doesn't
// read back the same way. `shown` is how many examples of each group to print
// (one by default).
//
// What commonmark.js does differently from the spec, and what the formatter
// does on purpose, is grouped under a name rather than shown: trailing
// whitespace on a line of an html block, `'` written as `&#x27;` in a
// destination, an extension of the formatter (a definition list, front matter),
// and whitespace outside of code, which the formatter decides for itself.

interface Example {
  markdown: string;
  formatted: string | undefined;
}

const [path, kind, shownText = "1"] = Deno.args;
if (!path || (kind !== "parse" && kind !== "format")) {
  console.error("usage: classify.ts <report.txt> parse|format [shown]");
  Deno.exit(1);
}
const shown = Number(shownText);
const entries = (await Deno.readTextFile(path)).split(/^==== /m).slice(1);
const groups = new Map<string, { expected: string; actual: string; examples: Example[] }>();
for (const entry of entries) {
  const head = entry.split("\n")[0];
  if (kind === "parse" && !head.includes("parse:FAIL")) {
    continue;
  }
  if (kind === "format" && !(head.includes("parse:ok") && head.includes("format:FAIL"))) {
    continue;
  }
  const { expected, actual, example } = classify(entry, kind);
  const key = JSON.stringify([expected, actual]);
  const group = groups.get(key) ?? { expected, actual, examples: [] };
  group.examples.push(example);
  groups.set(key, group);
}

const failures = [...groups.values()].reduce((count, group) => count + group.examples.length, 0);
console.log(`${entries.length} entries, ${failures} of them ${kind} failures in ${groups.size} groups`);
for (const group of [...groups.values()].sort((a, b) => b.examples.length - a.examples.length)) {
  console.log(
    `==== ${group.examples.length} | expected: ${JSON.stringify(group.expected)} | actual: ${
      JSON.stringify(group.actual)
    }`,
  );
  if (!group.actual) {
    continue; // a named group, which there is nothing to look at in
  }
  for (const example of group.examples.slice(0, shown)) {
    console.log(`   markdown:  ${JSON.stringify(example.markdown)}`);
    if (example.formatted !== undefined) {
      console.log(`   formatted: ${JSON.stringify(example.formatted)}`);
    }
  }
}

function classify(entry: string, kind: string): { expected: string; actual: string; example: Example } {
  const markdown = field(entry, "markdown", "expected") ?? "";
  let expected = field(entry, "expected", "parsed|formatted markdown");
  let actual;
  let formatted;
  if (kind === "parse") {
    actual = field(entry, "parsed", "formatted markdown");
  } else {
    actual = field(entry, "formatted html", "note");
    formatted = field(entry, "formatted markdown", "formatted html");
  }
  const example = { markdown, formatted };
  if (expected === undefined || actual === undefined) {
    return { expected: "<incomplete entry>", actual: "", example };
  }
  if (actual.includes("extension:")) {
    return { expected: "extension of the formatter", actual: "", example };
  }
  expected = expected.replaceAll("&#x27;", "'");
  actual = actual.replaceAll("&#x27;", "'");
  let expectedParts;
  let actualParts;
  if (kind === "format") {
    expected = loosen(expected);
    actual = loosen(actual);
    if (expected === actual) {
      return { expected: "whitespace only", actual: "", example };
    }
    expectedParts = expected.split(/(?=<)/);
    actualParts = actual.split(/(?=<)/);
  } else {
    if (stripLines(expected) === stripLines(actual)) {
      return { expected: "trailing whitespace only", actual: "", example };
    }
    expectedParts = expected.split("\n");
    actualParts = actual.split("\n");
  }
  for (let i = 0; i < Math.min(expectedParts.length, actualParts.length); i++) {
    if (expectedParts[i] !== actualParts[i]) {
      return { expected: expectedParts[i].slice(0, 70), actual: actualParts[i].slice(0, 70), example };
    }
  }
  return { expected: "<length differs>", actual: String(expectedParts.length - actualParts.length), example };
}

function field(entry: string, name: string, nextNames: string): string | undefined {
  const match = entry.match(new RegExp(`--- ${name}\\n([\\s\\S]*?)(?=--- (?:${nextNames})\\n|$)`));
  return match?.[1];
}

function stripLines(text: string): string {
  return text.split("\n").map((line) => line.replace(/[ \t]+$/, "")).join("\n");
}

/** The html with the whitespace outside of code collapsed, the way the spec check compares formatted output. */
function loosen(html: string): string {
  let out = "";
  let rest = html;
  while (true) {
    const start = rest.indexOf("<pre>");
    if (start < 0) {
      return out + collapse(rest);
    }
    out += collapse(rest.slice(0, start));
    const close = rest.indexOf("</pre>", start);
    const end = close < 0 ? rest.length : close + 6;
    out += stripLines(rest.slice(start, end));
    rest = rest.slice(end);
  }
}

function collapse(html: string): string {
  let out = "";
  let inSpace = false;
  for (const c of html) {
    if (/\s/.test(c)) {
      inSpace = true;
      continue;
    }
    if (inSpace && !out.endsWith(">") && c !== "<") {
      out += " ";
    }
    inSpace = false;
    out += c;
  }
  return out;
}

File: scripts/spec_fuzz/classify.py
#!/usr/bin/env python3
"""Groups the failures of a spec check report by what differs first, so that
thousands of them can be read as the handful of causes they have.

    classify.py <report.txt> parse|format [shown]

`parse` reads the examples whose parse differs from the expected html, `format`
the ones that parsed as expected but whose formatted output doesn't read back
the same way. `shown` is how many examples of each group to print (one by
default).

What commonmark.js does differently from the spec, and what the formatter does
on purpose, is grouped under a name rather than shown: trailing whitespace on a
line of an html block, `'` written as `&#x27;` in a destination, an extension
of the formatter (a definition list, front matter), and whitespace outside of
code, which the formatter decides for itself.
"""
import collections
import re
import sys


def main():
    path, kind = sys.argv[1], sys.argv[2]
    shown = int(sys.argv[3]) if len(sys.argv) > 3 else 1
    entries = re.split(r"^==== ", open(path, encoding="utf-8").read(), flags=re.M)[1:]
    groups = collections.OrderedDict()
    for entry in entries:
        head = entry.split("\n")[0]
        if kind == "parse" and "parse:FAIL" not in head:
            continue
        if kind == "format" and ("parse:ok" not in head or "format:FAIL" not in head):
            continue
        key, example = classify(entry, kind)
        groups.setdefault(key, []).append(example)

    print(len(entries), "entries,", sum(len(g) for g in groups.values()), "of them", kind, "failures in", len(groups), "groups")
    for key, examples in sorted(groups.items(), key=lambda pair: -len(pair[1])):
        print("====", len(examples), "| expected:", repr(key[0]), "| actual:", repr(key[1]))
        if not key[1]:
            continue  # a named group, which there is nothing to look at in
        for markdown, formatted in examples[:shown]:
            print("   markdown: ", repr(markdown))
            if formatted is not None:
                print("   formatted:", repr(formatted))


def classify(entry, kind):
    markdown = field(entry, "markdown", "expected")
    expected = field(entry, "expected", "parsed|formatted markdown")
    if kind == "parse":
        actual = field(entry, "parsed", "formatted markdown")
        formatted = None
    else:
        actual = field(entry, "formatted html", "note")
        formatted = field(entry, "formatted markdown", "formatted html")
    example = (markdown, formatted)
    if actual is None or expected is None:
        return ("<incomplete entry>", ""), example
    if "extension:" in actual:
        return ("extension of the formatter", ""), example
    expected, actual = expected.replace("&#x27;", "'"), actual.replace("&#x27;", "'")
    if kind == "format":
        expected, actual = loosen(expected), loosen(actual)
        if expected == actual:
            return ("whitespace only", ""), example
        expected_parts, actual_parts = re.split(r"(?=<)", expected), re.split(r"(?=<)", actual)
    else:
        if strip_lines(expected) == strip_lines(actual):
            return ("trailing whitespace only", ""), example
        expected_parts, actual_parts = expected.split("\n"), actual.split("\n")
    for expected_part, actual_part in zip(expected_parts, actual_parts):
        if expected_part != actual_part:
            return (expected_part[:70], actual_part[:70]), example
    return ("<length differs>", str(len(expected_parts) - len(actual_parts))), example


def field(entry, name, next_names):
    match = re.search(r"--- " + name + r"\n(.*?)(?=--- (?:" + next_names + r")\n|\Z)", entry, re.S)
    return match.group(1) if match else None


def strip_lines(text):
    return "\n".join(line.rstrip(" \t") for line in text.split("\n"))


def loosen(html):
    """The html with the whitespace outside of code collapsed, the way the spec
    check compares formatted output."""
    out = ""
    rest = html
    while True:
        start = rest.find("<pre>")
        if start < 0:
            return out + collapse(rest)
        out += collapse(rest[:start])
        end = rest.find("</pre>", start)
        end = len(rest) if end < 0 else end + 6
        out += strip_lines(rest[start:end])
        rest = rest[end:]


def collapse(html):
    out = []
    in_space = False
    for c in html:
        if c.isspace():
            in_space = True
            continue
        if in_space and not (out and out[-1] == ">") and c != "<":
            out.append(" ")
        in_space = False
        out.append(c)
    return "".join(out)


if __name__ == "__main__":
    main()
